state/store.go

```diff
@@ -212,133 +234,189 @@ func (store dbStore) Bootstrap(state State) error {
+		return err
+	}
+
+	return nil
+}
+
+func (store dbStore) pruneValidatorSets(height int64) error {
+	valInfo, err := loadValidatorsInfo(store.db, height)
 	if err != nil {
-		return fmt.Errorf("consensus params at height %v not found: %w", to, err)
+		return fmt.Errorf("validators at height %v not found: %w", height, err)
 	}
 
-	keepVals := make(map[int64]bool)
+	var (
+		lastRecordedValSetHeight int64
+		lastRecordedValSet       *tmstate.ValidatorsInfo
+	)
+
+	// We will prune up to the validator set at the given "height". As we don't save validator sets every
+	// height but only when they change or at a check point, it is likely that the validator set at the height
+	// we prune to is empty and thus dependent on the validator set saved at a previous height. We must find
+	// that validator set and make sure it is not pruned by saving after we have finished pruning.
 	if valInfo.ValidatorSet == nil {
-		keepVals[valInfo.LastHeightChanged] = true
-		keepVals[lastStoredHeightFor(to, valInfo.LastHeightChanged)] = true // keep last checkpoint too
+		lastRecordedValSetHeight = lastStoredHeightFor(height, valInfo.LastHeightChanged)
+		lastRecordedValSet, err = loadValidatorsInfo(store.db, lastRecordedValSetHeight)
+		if err != nil || lastRecordedValSet.ValidatorSet == nil {
+			return fmt.Errorf("couldn't find validators at height %d (height %d was originally requested): %w",
+				lastStoredHeightFor(height, valInfo.LastHeightChanged),
+				height,
+				err,
+			)
+		}
 	}
-	keepParams := make(map[int64]bool)
-	if paramsInfo.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
-		keepParams[paramsInfo.LastHeightChanged] = true
+
+	// batch delete all the validators sets up to height
+	if err := store.batchDelete(validatorsKey, height); err != nil {
+		return err
 	}
 
-	batch := store.db.NewBatch()
-	defer batch.Close()
-	pruned := uint64(0)
-
-	// We have to delete in reverse order, to avoid deleting previous heights that have validator
-	// sets and consensus params that we may need to retrieve.
-	for h := to - 1; h >= from; h-- {
-		// For heights we keep, we must make sure they have the full validator set or consensus
-		// params, otherwise they will panic if they're retrieved directly (instead of
-		// indirectly via a LastHeightChanged pointer).
-		if keepVals[h] {
-			v, err := loadValidatorsInfo(store.db, h)
-			if err != nil || v.ValidatorSet == nil {
-				vip, err := store.LoadValidators(h)
-				if err != nil {
-					return err
-				}
-
-				pvi, err := vip.ToProto()
-				if err != nil {
-					return err
-				}
-
-				v.ValidatorSet = pvi
-				v.LastHeightChanged = h
-
-				bz, err := v.Marshal()
-				if err != nil {
-					return err
-				}
-				err = batch.Set(calcValidatorsKey(h), bz)
-				if err != nil {
-					return err
-				}
-			}
-		} else {
-			err = batch.Delete(calcValidatorsKey(h))
-			if err != nil {
-				return err
-			}
+	// now we recover the last recorded validator set if it had been set earlier
```

Comment:
This isn't crash-safe. If the node crashes after (or during) pruning and doesn't get around to recreating the key, the database is in an inconsistent and unusable state. You could write the key in the same batch as the deletions (although not all databases are guaranteed to write batches atomically), but it would be much better to simply not delete the keys in the first place.

Reply:
Also realized that this may break RPC validator set queries until the entry has been recreated.

Reply:
Hmm even if I don't delete the dependent keys and just skip over them it could still be that the database crashes at a point before reaching the dependent key in which case it is also in an unusable state. It will probably be better if I revert to the prior pattern of deleting them in order of highest height to lowest height.